c++/src/kj/common.h

```diff
@@ -1928,6 +1928,30 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     return true;
   }
 
+  inline bool operator<(const ArrayPtr& other) const {
+    size_t comparisonSize = kj::min(size_, other.size_);
+    if constexpr (isIntegral<RemoveConst<T>>()) {
```

Comment:
For multi-byte integers on little-endian architectures, `memcmp()` will not produce the same result the for loop, so we have to check that the size is 1:

```suggestion
    if constexpr (isIntegral<RemoveConst<T>>() && sizeof(T) == 1) {
```

But actually, even that might be wrong if the elements are signed, as memcmp() will treat them as unsigned.

But `char` (with no qualifiers) is signed on some systems and not on others, but presumably "correct" sorting of strings should always treat it as unsigned?

So maybe we actually want to use the `memcmp()` approach specifically for `char` and `unsigned char`, but not any other type?

Reply:
Please also add tests comparing arrays of `int`, covering negative values and values larger than 256.

Reply:
`strcmp()` is explicitly documented (on the man page) as presuming unsigned characters, which seems desirable.

It's sort of weird if we say comparing ArrayPtr<char> will not produce the same results as a loop comparing individual values, on platforms where char is signed. However, I think for all real-world purposes, this is what people will want. If they actually want to compare signed chars they can use `ArrayPtr<signed char>`.

Reply:
(That is to say, for `ArrayPtr<char>` or `ArrayPtr<unsigned char>`, using `memcmp()` seems like the right thing. But for `signed char` or multi-byte integers it is still wrong.)

Reply:
- Is our goal then to support little endian architectures?
  - If so, I think it makes sense to add your change.
Should we specialize to `strcmp` for `signed char`?
  - This is something that would be known at compile time, so it makes sense to me to do this if we truly want to specialize.
[ ] Add test comparing arrays of ints including negative values and values larger than 256.

Reply:
> Is our goal then to support little endian architectures?

All modern architectures are little-endian.

> If so, I think we should detect the target's endianness at compile-time right?

Not in this case. While technically memcmp() would work correctly for unsigned integers on big-endian architectures, there are no big-endian architectures anymore, so writing such code would be pointless. We should just not use memcmp() for multi-byte integers.

Reply:
> All modern architectures are little-endian.

This was the missing link for me.

Reply:
Should be fixed now.